internal/channel/channel.go

```diff
@@ -20,109 +18,145 @@ type Channel struct {
+	if err != nil {
+		c.Logger.Errorw("Failed to initialize channel plugin", zap.Error(err))
 		return nil
 	}
 
-	if matched, _ := regexp.MatchString("^[a-zA-Z0-9]*$", c.Type); !matched {
-		return fmt.Errorf("channel type must only contain a-zA-Z0-9, %q given", c.Type)
+	if err := p.SetConfig(c.Config); err != nil {
+		c.Logger.Errorw("Failed to set channel plugin config, terminating the plugin", zap.Error(err))
+		p.Stop()
+		return nil
 	}
 
-	path := filepath.Join(pluginDir, c.Type)
+	return p
+}
 
-	cmd := exec.Command(path)
+// runPlugin is called as go routine to initialize and maintain the plugin by receiving signals on given chan(s)
+func (c *Channel) runPlugin() {
+	var currentlyRunningPlugin *Plugin
+	for {
+		if currentlyRunningPlugin == nil {
+			currentlyRunningPlugin = c.initPlugin()
+		}
 
-	writer, err := cmd.StdinPipe()
-	if err != nil {
-		return fmt.Errorf("failed to create stdin pipe: %w", err)
-	}
+		var rpcDone <-chan struct{}
+		if currentlyRunningPlugin != nil {
+			rpcDone = currentlyRunningPlugin.rpc.Done()
+		}
 
-	reader, err := cmd.StdoutPipe()
-	if err != nil {
-		return fmt.Errorf("failed to create stdout pipe: %w", err)
-	}
+		select {
+		case <-rpcDone:
+			c.Logger.Debug("Receive plugin error signal")
 
-	errPipe, err := cmd.StderrPipe()
-	if err != nil {
-		return fmt.Errorf("failed to create stderr pipe: %w", err)
-	}
+			if currentlyRunningPlugin != nil {
+				currentlyRunningPlugin.Stop()
+				currentlyRunningPlugin = nil
+			}
 
-	go forwardLogs(errPipe, c.Logger)
+			continue
+		case <-c.newConfigCh:
+			c.Logger.Debug("Receive reload plugin config signal")
 
-	if err = cmd.Start(); err != nil {
-		return fmt.Errorf("failed to start cmd: %w", err)
-	}
-	c.Logger.Debug("Cmd started successfully")
+			if currentlyRunningPlugin != nil {
+				currentlyRunningPlugin.Stop()
+				currentlyRunningPlugin = nil
+			}
 
-	c.cmd = cmd
-	c.rpc = rpc.NewRPC(writer, reader, c.Logger)
+			continue
+		case <-c.stopPluginCh:
+			c.Logger.Debug("Receive plugin stop signal")
 
-	if err = c.SetConfig(c.Config); err != nil {
-		go c.terminate(c.cmd, c.rpc)
+			if currentlyRunningPlugin != nil {
+				currentlyRunningPlugin.Stop()
+				currentlyRunningPlugin = nil
+			}
 
-		c.cmd = nil
-		c.rpc = nil
+			return
+		case req := <-c.notificationCh:
+			if currentlyRunningPlugin == nil {
+				currentlyRunningPlugin = c.initPlugin()
+			}
 
-		return fmt.Errorf("failed to set config: %w", err)
+			if currentlyRunningPlugin == nil {
+				c.Logger.Debug("Cannot send notification, plugin could not be started")
+				req.SetErr(errors.New("plugin could not be started"))
+			} else {
+				go func(p *Plugin) {
+					p.SendNotification(req)
+				}(currentlyRunningPlugin)
+			}
+		}
 	}
-	c.Logger.Debugw("Successfully set config", zap.String("config", c.Config))
-
-	return nil
 }
 
+// Stop ends the lifecycle of its plugin.
+// This should only be called when the channel is not more required.
 func (c *Channel) Stop() {
```

Comment:
Still doesn't say if it's safe/intended to be called multiple times.

Reply:
Yeah, it's not safe. And it should never call multiple times.